### ISSUE TITLE
Clarification between tagged_na() and user-defined missing values

### ISSUE THREAD

**Issue by larmarange** [larmarange/labelled#93]
@dusadrian

thank you for your different comments. However, I have the feeling that some clarifications could help.

### SPSS's style user-defined missing values

`na_range()` and `na_values()` are intended to indicate some real values that should be considered as missing values. Therefore, there is no sense to assign a NA or event a tagged NA through `na_range()`. It should always be a **real value**.

Such verification is currently not done in `haven::labelled_spss()` and I would suggest opening a ticket for `haven` to be more strict when defining `labelled_spss()`. Such verification should also be added in `labelled`. 

Now a more fundamental question is how such values should be treated and I agree that there are some inconsistencies. From my personal point of view, it indicates that these values **should/could** be considered as missing values but not that they have to be treated as missing values. Remembering that labelled vectors are a sort of intermediate format, it should be up to the users to indicate if and when these user-defined missing values should be treated as missing values, by converting them to explicit missing values with `user_na_to_na()`.

It is TRUE that the fact that `is.na()` returns NA in such a case is confusing. It was a choice of `haven` and in the philosophy of `haven`, labelled vectors are just an intermediate format to facilitate import and these vectors should be converted into other classic R classes.

It would have been better in `is.na.labelled_spss()` to have added an option to let the user decide if user_na should be considered as NA or not. But this method does not allow any additional argument.

We could add in `labelled` a more generic `is_na()` method whith an option user_na_to_na (FALSE or TRUE) to indicated how user NAs should be treated, as well as a `is_user_na()` function.

Developing a dedicated vignette about how to correctly use NA values with `labelled` could be relevant.

### Tagged NAs

This feature was also introduced by `haven` to reproduce the way Stata or SAS works. It has never been intended to be used with SPSS'style user NAs.

However, I can understand the need of associating a value label to such tagged NAs (but not to use them with `na_range()`).

### Switching from tagged NAs to SPSS's user na

Maybe it could be relevant to add in `labelled` two new functions `user_na_to_tagged_na()` and `tagged_na_to_user_na()` to facilitate the switch from one system to the other.

**Comment by larmarange** [comment]
See #94 for implementation

**Comment by dusadrian** [comment]
Thank you very much for the detailed explanation, much appreciated.

As also mentioned in issue [#94](https://github.com/larmarange/labelled/pull/94), I agree that `haven` introduced tagged NAs "to reproduce the way Stata or SAS works", but I am somewhat baffled as to why "it has never been intended to be used with SPSS'style user NAs."

Stata and SAS users routinely use such tagged values to differentiate between different missing values. Which means these are user defined, much in the same spirit that SPSS users define missing values using real numbers. The end goal is the same, only the methods differ.

To put it otherwise, what is the purpose (in R) of a tagged NA if not to allow users to define a missing value?

Quite the contrary, I believe tagged NAs are very useful (but still user defined) missing values, and as such they should be treated exactly the same way as the other (SPSS style) user defined missing values with real numbers. 

Whatever Hadley's original reason was, tagged NAs will universally be used as user defined missing values anyways. And if we could hopefully agree on this point, then packages `labelled` and `haven` should solve the inconsistencies between
SPSS (user defined) missing values and tagged (user defined) missing values.

I know this originates in package `haven` of course, but Hadley is a co-author here too and the link between these two packages is very strong (many of the `haven` functions are directly imported by package `labelled`)

**Comment by larmarange** [comment]
There is a fundamental difference:

- tagged NAs are missing values (missing values in data)
- SPSS's user NAs are normal values, with metadata indicating that they could be interpreted as NAs, letting the user decide to convert them into proper NAs or not

This is exactly the same thing regarding a categorical variable coded with numeric values and value labels. From R perspective, this is not a categorical variable as the proper way in R to code categorical variables are factors. A labelled vector remains a numeric vector with additional metadata (the value labels) and users **must** convert such vectors into proper factors if they want to analyze them as categorical variables, for example in a regression model or for plotting.

**Comment by dusadrian** [comment]
I agree with the difference between labels and factors, it is the users' decision to convert to proper factors when using as categorical variables.

However, it is simply not true that SPSS "let users decide" to convert  into proper NAs or not. Quite the contrary, once a certain value was declared as missing, SPSS users don't (have to) do anything else at all.

Calculate a mean? No problem, declared missing values are not taken into account.
Use a categorical variable in a regression model? No problem, the categories associated with declared missing values are not taken into account.

Exactly the same thing happens in Stata and SAS. Once a value has been declared as "missing", they are not taken into account. The emphasis, here, is not on the difference between a tagged NA and a number, but on the fact that all software **declare** missing values **and nothing else** should be done further.

R, on the other hand, seems to be different. Just to clarify, I am not complaining for myself (I do just fine with regular factors): but I am trying to convert many of my faculty peers to R, and if such mundane things in other software become so clumsy in R, there is little chance to convince. And this is a pity, because the very point of package `haven`, when introducing these declared missing values, was to make foreign people more at home. The current behaviour simply defeats the purpose.

**Comment by larmarange** [comment]
**R** is not **SPSS**!!!

The proper way to deal with missing values in **R** is to use `NA`. This is expected in all statistical functions in R. You can eventually use tagged NAs are there still internally considered as `NA` in all R functions.

The purpose of `labelled_spss()` is **NOT** to reproduce the way SPSS deals with missing values. This is a **transitional** representation of data when imported from SPSS to allow users to keep the detailed information about the different kinds of missing values as defined in SPSS. However, for analysis, users **must** decide to convert them into proper `NA`s or to convert them into proper values. 

The proper to do statistics should therefore be:

``` r
library(labelled)

x <- labelled_spss(c(1, 1, 2, 3, 9, 2, 9, 3), na_values = 9)

mean(user_na_to_na(x), na.rm = TRUE)
#> [1] 2

mean(remove_user_na(x))
#> [1] 3.75
```

<sup>Created on 2021-04-06 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by dusadrian** [comment]
That makes a lot of sense. Whatever these objects are, in the end they still have to obey R's general way of dealing with NA values. It is a very small price to pay for the overall superior functionality, and this is a very instructive thread for anyone interested in this topic.

Just to make sure I understand your example right, there is absolutely no difference between `mean(x)` and `mean(remove_user_na(x))`, because `remove_user_na()` doesn't necessarily remove missing values: its primary purpose is to remove the metadata about those missing values.

**Comment by dusadrian** [comment]
@larmarange see tidyverse/haven#574 for a counterexample.
Actually, it is possible to have a real value to be treated as missing.
